tests/unit/apps/test_base.py

```diff
@@ -0,0 +1,240 @@
+    app = OpenStackApplication(
+        "gnocchi",
+        status["gnocchi_ussuri"],
+        app_config,
+        model,
+        "gnocchi",
+    )
+    assert app.is_os_channel_based is True
+    assert app.current_os_release == "ussuri"
+
+
+def test_application_designate_bind_ussuri(status, config, model):
+    app_config = config["openstack_ussuri"]
+    app_config["action-managed-upgrade"] = {"value": False}
+    app = OpenStackApplication(
+        "designate-bind",
+        status["designate_bind_ussuri"],
+        app_config,
+        model,
+        "designate-bind",
+    )
+    assert app.is_os_channel_based is True
+    assert app.current_os_release == "ussuri"
+
+
+def test_application_gnocchi_upgrade_plan_ussuri_to_victoria(status, config, model):
+    target = OpenStackRelease("victoria")
+    app_config = config["openstack_ussuri"]
+    app_config["action-managed-upgrade"] = {"value": False}
+    app = OpenStackApplication(
+        "gnocchi",
+        status["gnocchi_ussuri"],
+        app_config,
+        model,
+        "gnocchi",
+    )
+
+    upgrade_plan = app.generate_upgrade_plan(target)
+
+    expected_plan = UpgradeStep(
+        description=f"Upgrade plan for '{app.name}' to {target}",
+        parallel=False,
+    )
+
+    upgrade_steps = [
+        UpgradeStep(
+            description=(
+                f"Upgrade software packages of '{app.name}' from the current APT repositories"
+            ),
+            parallel=False,
+            coro=app_utils.upgrade_packages(app.status.units.keys(), model, None),
+        ),
+        UpgradeStep(
+            description=f"Refresh '{app.name}' to the latest revision of 'ussuri/stable'",
+            parallel=False,
+            coro=model.upgrade_charm(app.name, "ussuri/stable", switch=None),
+        ),
+        UpgradeStep(
+            description=f"Upgrade '{app.name}' to the new channel: 'victoria/stable'",
+            parallel=False,
+            coro=model.upgrade_charm(app.name, "victoria/stable"),
+        ),
+        UpgradeStep(
+            description=(
+                f"Change charm config of '{app.name}' "
+                f"'{app.origin_setting}' to 'cloud:focal-victoria'"
+            ),
+            parallel=False,
+            coro=model.set_application_config(
+                app.name,
+                {f"{app.origin_setting}": "cloud:focal-victoria"},
+            ),
+        ),
+        UpgradeStep(
+            description=f"Wait 120 s for app {app.name} to reach the idle state.",
+            parallel=False,
+            coro=model.wait_for_idle(120, [app.name]),
+        ),
+        UpgradeStep(
+            description=f"Check if the workload of '{app.name}' has been upgraded",
+            parallel=False,
+            coro=app._check_upgrade(target),
+        ),
+    ]
+
+    add_steps(expected_plan, upgrade_steps)
+
+    assert upgrade_plan == expected_plan
+
+
+def test_application_gnocchi_upgrade_plan_xena_to_yoga(status, config, model):
```

Comment:
By looking at the test code, it's unclear why we would need to test upgrade plan for both ussuri->victoria and xena->yoga. It there are additional things covered by one of the plan, we can mention it in the docstring? Otherwise, I think we should remove one of them.